browser/onboarding/onboarding_tab_helper.cc

```diff
@@ -0,0 +1,113 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/onboarding/onboarding_tab_helper.h"
+
+#include <string>
+#include <vector>
+
+#include "base/strings/string_number_conversions.h"
+#include "brave/browser/onboarding/domain_map.h"
+#include "brave/browser/onboarding/pref_names.h"
+#include "brave/browser/ui/views/frame/brave_browser_view.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/first_run/first_run.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "components/grit/brave_components_strings.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace onboarding {
+
+OnboardingTabHelper::~OnboardingTabHelper() = default;
+
+OnboardingTabHelper::OnboardingTabHelper(content::WebContents* web_contents)
+    : content::WebContentsObserver(web_contents),
+      content::WebContentsUserData<OnboardingTabHelper>(*web_contents) {}
+
+void OnboardingTabHelper::DidStopLoading() {
+  Browser* browser = chrome::FindBrowserWithWebContents(web_contents());
+  auto* shields_data_controller =
+      brave_shields::BraveShieldsDataController::FromWebContents(
+          web_contents());
+  auto* active_tab_web_contents =
+      browser->tab_strip_model()->GetActiveWebContents();
+
+  if (web_contents() != active_tab_web_contents)
+    return;
+
+  if (shields_data_controller &&
```

Comment:
DCHECK on shields_data_controller?

Reply:
> What if the web_contents() is null

I feel like it's unlikely to happen because we're inside the callback from it. But if you think it could happen, we can also add DCHECK(web_contents()); in on the top of this function.